### PR TITLE
feat: add input component as types, simplify event handlers - emailpassword recipe 

### PR DESCRIPTION
**Pull request by amitbadala**
## Summary of change

Add existing prop **inputComponent** to be used to add Custom Components

## Related issues

-   [Field decorators #36](https://github.com/supertokens/supertokens-auth-react/issues/36)

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [x] Changelog has been updated
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [x] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [x] Had run `npm run build-pretty`
-   [x] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [x] If added a new recipe interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [x] If I added a new recipe, I also added the recipe entry point into the `size-limit` section of `package.json` with the size limit set to the current size rounded up.
-   [x] If I added a new recipe, I also added the recipe entry point into the `rollup.config.mjs`

### CODE REVIEW

File: lib/ts/types.ts [hunk 1]
```diff
@@ -283,6 +289,11 @@ export type NormalisedFormField = {
      * Moves focus to the input element when it mounts
      */
     autofocus?: boolean;
+
+    /*
+     * Ability to add custom components
+     */
+    inputComponent?: React.FC<InputProps>;
```
Comment (rishabhpoddar) [inline lib/ts/types.ts:296]: decide if this needs to be here or not

File: examples/for-tests/src/App.js [hunk 2]
```diff
@@ -168,6 +168,52 @@ const formFields = [
...
+                }}>
+                <input
+                    name={name}
+                    type="checkbox"
+                    onChange={(e) => rest.onChange({ id: name, value: e.target.checked })}></input>
+                <span style={{ marginLeft: 5 }}>I agree to the terms and conditions</span>
+            </div>
+        ),
+        validate: async (value) => {
+            if (value) return undefined;
```
Comment (rishabhpoddar) [inline examples/for-tests/src/App.js:211]:
```suggestion
            if (value === undefined || value === null || value === "") {
                return undefined;
            }
```

File: test/end-to-end/signup.test.js [hunk 3]
```diff
@@ -414,3 +415,134 @@ describe("SuperTokens SignUp => Server Error", function () {
...
+    //     await browser.close();
+    // });
+
+    afterEach(function () {
+        return screenshotOnFailure(this, browser);
+    });
+
+    beforeEach(async function () {
+        consoleLogs = [];
+        consoleLogs = await clearBrowserCookiesWithoutAffectingConsole(page, consoleLogs);
```
Comment (rishabhpoddar) [inline test/end-to-end/signup.test.js:450]: given that we are adding this in the top level describe, which already has this, do we also need to do this here again?

File: test/end-to-end/signup.test.js [hunk 4]
```diff
@@ -414,3 +415,134 @@ describe("SuperTokens SignUp => Server Error", function () {
...
+        await submitForm(page);
+        formFieldErrors = await getFieldErrors(page);
+        assert.deepStrictEqual(formFieldErrors, []);
+    });
+
+    it.only("Check if custom values are part of the signup payload", async function () {
+        const customFieldNames = ["terms", "ratings"];
+        const requestHandler = (request) => {
+            if (request.url().includes(SIGN_UP_API) && request.method() === "POST") {
+                assert.ok(
```
[the remark this answers is not in the document]
Comment (rishabhpoddar) [reply test/end-to-end/signup.test.js:513]: Also, are you sure that the execution is even coming in the interceptor?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: test/end-to-end/signup.test.js [hunk 5]
```diff
@@ -414,3 +415,134 @@ describe("SuperTokens SignUp => Server Error", function () {
...
+        // check if select dropdown is loaded
+        const selectDropdownExists = await getCustomComponents(page, "select");
+        assert.ok(selectDropdownExists, "Select dropdown exists");
+
+        // check if checbox is loaded
+        const checkboxExists = await getCustomComponents(page, 'input[type="checkbox"]');
+        assert.ok(checkboxExists, "Checkbox exists");
+    });
+
+    it("Should show error messages, based on the validation", async function () {
```
[the remark this answers is not in the document]
Comment (amitbadala) [reply test/end-to-end/signup.test.js]: updated!

File: test/end-to-end/signup.test.js (hunk 3 above)
Comment (amitbadala) [reply test/end-to-end/signup.test.js]: now, using the top-level hooks

File: test/end-to-end/signup.test.js [hunk 6]
```diff
@@ -414,3 +415,134 @@ describe("SuperTokens SignUp => Server Error", function () {
...
+
+    beforeEach(async function () {
+        consoleLogs = [];
+        consoleLogs = await clearBrowserCookiesWithoutAffectingConsole(page, consoleLogs);
+        await Promise.all([
+            page.goto(`${TEST_CLIENT_BASE_URL}`),
+            page.waitForNavigation({ waitUntil: "networkidle0" }),
+        ]);
+
+        await page.evaluate(() => window.localStorage.setItem("SHOW_CUSTOM_FIELDS", "YES"));
```
[the remark this answers is not in the document]
Comment (amitbadala) [reply test/end-to-end/signup.test.js]: yup, it fails as expected

File: lib/ts/recipe/emailpassword/types.ts [hunk 7]
```diff
@@ -225,15 +223,11 @@ export type NormalisedEnterEmailForm = FeatureBaseConfig & {
...
 };
 
-/*
- * Props Types.
- */
-
 type FormThemeBaseProps = ThemeBaseProps & {
     /*
-     * Form fields to use in the signin form.
+     * Form fields to use in the signup form.
```
Comment (rishabhpoddar) [inline lib/ts/recipe/emailpassword/types.ts:228]: add a comment explaining why omit

File: lib/ts/recipe/emailpassword/types.ts [hunk 8]
```diff
@@ -225,15 +223,11 @@ export type NormalisedEnterEmailForm = FeatureBaseConfig & {
     formFields: NormalisedFormField[];
 };
 
-/*
- * Props Types.
- */
-
 type FormThemeBaseProps = ThemeBaseProps & {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/emailpassword/types.ts:226]:
```suggestion
type NonSignUpFormThemeBaseProps = ThemeBaseProps & {
```

File: lib/ts/recipe/emailpassword/types.ts [hunk 9]
```diff
@@ -288,6 +279,11 @@ export type FormFieldThemeProps = NormalisedFormField & {
      * Clears the field after calling the API.
      */
     clearOnSubmit?: boolean;
+
+    /*
+     * Ability to add custom components
+     */
+    inputComponent?: React.FC<InputProps>;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/emailpassword/types.ts:286]: the type of this does not have proper input. It says any

File: lib/ts/recipe/emailpassword/types.ts [hunk 10]
```diff
@@ -288,6 +279,11 @@ export type FormFieldThemeProps = NormalisedFormField & {
...
 };
```
Comment (rishabhpoddar) [inline lib/ts/recipe/emailpassword/types.ts:288]: modify the with-typescript folder in test to also use this new feature.

File: examples/for-tests/src/App.js [hunk 11]
```diff
@@ -168,6 +168,54 @@ const formFields = [
...
+const customFields = [
+    {
+        id: "ratings",
+        label: "Ratings",
+        inputComponent: ({ value, name, ...rest }) => (
+            <select
+                name={name}
+                data-supertokens="inputComponent"
+                placeholder="Add Ratings"
+                onChange={(e) => rest.onChange({ id: name, value: e.target.value })}>
```
Comment (rishabhpoddar) [inline examples/for-tests/src/App.js:180]: understand how onChange works:
- Does it require id and value?
- Can value be anything other than a string?
- If id is required, can be get rid of it?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: examples/for-tests/src/App.js (hunk 11 above)
Comment (amitbadala) [reply examples/for-tests/src/App.js:180]: Does it require id and value? -` Yes, both are required`
Can value be anything other than a string? - `No`
If id is required, can be get rid of it? - `Need to write a separate onChange func to handle inputComponent onChange`

File: lib/ts/recipe/passwordless/components/themes/signInUp/phoneNumberInput.tsx [hunk 12]
```diff
@@ -71,10 +71,7 @@ function PhoneNumberInput({
     const handleChange = useCallback(
         (newValue: string) => {
             if (onChangeRef.current !== undefined) {
-                onChangeRef.current({
-                    id: name,
-                    value: newValue,
-                });
+                onChangeRef.current(newValue);
```
Comment (rishabhpoddar) [inline lib/ts/recipe/passwordless/components/themes/signInUp/phoneNumberInput.tsx:74]: Is this change detected cause of a TS error or something else?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/emailpassword/components/library/formBase.tsx [hunk 13]
```diff
@@ -219,7 +246,7 @@ export const FormBase: React.FC<FormBaseProps<any>> = (props) => {
                                     autofocus={field.autofocus}
                                     onInputFocus={onInputFocus}
                                     onInputBlur={onInputBlur}
```
Comment (rishabhpoddar) [inline lib/ts/recipe/emailpassword/components/library/formBase.tsx:248]: change above const fstate part to:
```
const fstate: FieldState | undefined = fieldStates.find((s) => s.id === field.id);
                if (fstate === undefined) {
                    throw new Error("Should never come here");
                }
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/passwordless/components/themes/signInUp/phoneNumberInput.tsx (hunk 12 above)
Comment (amitbadala) [reply lib/ts/recipe/passwordless/components/themes/signInUp/phoneNumberInput.tsx:74]: yes

File: test/end-to-end/signup.test.js [hunk 14]
```diff
@@ -342,6 +344,151 @@ describe("SuperTokens SignUp", function () {
...
+                country: "USA",
+                ratings: "good",
+            };
+
+            await setInputValues(page, [{ name: "country", value: updatedFields["country"] }]);
+            await setSelectDropdownValue(page, 'select[name="ratings"]', updatedFields["ratings"]);
+
+            // input field default value
+            const countryInput = await getInputField(page, "country");
+            const defaultCountry = await countryInput.evaluate((f) => f.value);
```
Comment (rishabhpoddar) [inline test/end-to-end/signup.test.js:483]:
```suggestion
            const updatedCountry = await countryInput.evaluate((f) => f.value);
```

File: test/end-to-end/signup.test.js [hunk 15]
```diff
@@ -342,6 +344,151 @@ describe("SuperTokens SignUp", function () {
...
+            assert.strictEqual(defaultCountry, updatedFields["country"]);
+
+            // dropdown default value is also getting set correctly
+            const ratingsDropdown = await waitForSTElement(page, "select");
+            const defaultRating = await ratingsDropdown.evaluate((f) => f.value);
```
Comment (rishabhpoddar) [inline test/end-to-end/signup.test.js:488]:
```suggestion
            const updatedRating = await ratingsDropdown.evaluate((f) => f.value);
```

File: test/end-to-end/signup.test.js [hunk 16]
```diff
@@ -342,6 +344,151 @@ describe("SuperTokens SignUp", function () {
...
+            const countryInput = await getInputField(page, "country");
+            const defaultCountry = await countryInput.evaluate((f) => f.value);
+            assert.strictEqual(defaultCountry, fieldsWithDefault["country"]);
+
+            // custom dropdown default value is also getting set correctly
+            const ratingsDropdown = await waitForSTElement(page, "select");
+            const defaultRating = await ratingsDropdown.evaluate((f) => f.value);
+            assert.strictEqual(defaultRating, fieldsWithDefault["ratings"]);
+        });
+
```
Comment (rishabhpoddar) [inline test/end-to-end/signup.test.js:471]: add one more test in which we have the default values, and when you click on sign up without changing anything, those get sent to the API call.

File: test/with-typescript/src/App.tsx [hunk 17]
```diff
@@ -326,6 +326,97 @@ function getRecipeList() {
...
+                <option value="best">Best</option>
+            </select>
+        ),
+        optional: true,
+    },
+];
+
+const getFormFieldsForSignUp = () => {
+    return localStorage.getItem("SHOW_CUSTOM_FIELDS") === "YES" ? customFormFieldsWithDefault : formFieldsForSignUp;
+};
```
Comment (rishabhpoddar) [inline test/with-typescript/src/App.tsx:418]: no need to do all this. Just put all the form fields above in the init below

File: test/with-typescript/src/App.tsx [hunk 18]
```diff
@@ -326,6 +326,97 @@ function getRecipeList() {
...
+        inputComponent: (inputProps) => (
+            <select
+                value={inputProps.value}
+                name={inputProps.name}
+                onChange={(e) => {
+                    if (inputProps.onChange) {
+                        inputProps.onChange(e.target.value.toString());
+                    }
+                }}
+                placeholder="Add Ratings">
```
Comment (rishabhpoddar) [inline test/with-typescript/src/App.tsx:403]: also add onInputBlur, onInputFocus

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: test/helpers.js [hunk 19]
```diff
@@ -1021,3 +1020,8 @@ export async function backendBeforeEach() {
         }).catch(console.error);
     }
 }
+
+export async function getShadowRootHandle(page) {
```
Comment (rishabhpoddar) [inline test/helpers.js:1024]: see if the existing pattern of our code works and if not, why not

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 20]
```diff
@@ -5,6 +5,46 @@ All notable changes to this project will be documented in this file.
...
+                            <option value="option 2">Option 2</option>
+                            <option value="option 3">Option 3</option>
+                        </select>
+                    )
+                }
+            ]
+        }
+    }
+});
+...
```
Comment (rishabhpoddar) [inline CHANGELOG.md:45]:
```suggestion
```

File: CHANGELOG.md [hunk 21]
```diff
@@ -5,6 +5,46 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)
 
+## [0.36.0] - 2023-10-30
+
+### Added
+
+-   Introduced the capability to utilize custom components in the Email-Password based recipes' signup form fields by exposing inputComponent types.
+-   Implemented the functionality to assign default values to the form fields in the Email-Password based recipes.
+-   Enhanced the onChange function to operate independently without requiring an id field.
```
Comment (rishabhpoddar) [inline CHANGELOG.md:14]: this is unclear

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**